### PR TITLE
DBZ-1814 (binary handling mode)

### PR DESCRIPTION
**Pull request by gunnarmorling**
https://issues.redhat.com/browse/DBZ-1814

Supersedes https://github.com/debezium/debezium/pull/1516/.

@jpechane, could you review this one, in particular my amendments? Thanks!

### CODE REVIEW

**Comment by gunnarmorling** [comment]
@jpechane, just force-pushed to fix a test failure.

**Comment by gunnarmorling** [comment]
Rebased onto the DBZ-2033 branch, so to benefit from the test stablization done in that branch.

**Comment by hashhar** [comment]
@gunnarmorling I did look into the PostgreSQL specific code and it looks good.

I'll test this PR to see if addresses the concerns raised in https://issues.redhat.com/browse/DBZ-1266 but the tests look good and I think it should not be an issue.

Has a purely hypothetical question though, the current switching on BinaryHandlingMode doesn't allow for custom implementations but it shouldn't be an issue since I personally can't think of any other conversion I would've done apart from the ones already implemented.  
Custom may be useful if somebody has added their own custom data types into PostgreSQL and can implement a matching Java type to return as but I really hope nobody went to those lengths.

**Comment by jpechane** [comment]
Reviewing, unfortunately this PR now requires rebase due to docs conflicts

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 1]
```diff
@@ -1113,15 +1113,15 @@ The _semantic type_ describes how the Kafka Connect schema captures the _meaning
...
 |`io.debezium.time.Interval` +
 (when `interval.handling.mode` is set to `string`)
 |The string representation of the interval value that follows pattern `P<years>Y<months>M<days>DT<hours>H<minutes>M<seconds>S`, e.g. `P1Y2M3DT4H5M6.78S`
 
 |`BYTEA`
-|`BYTES`
+|`BYTES` or `STRING`
 |n/a
-|
+|Either the raw bytes (the default), a base64-encoded String, or a hex-encoded String, based on the {link-prefix}:{link-postgresql-connector}#connector-property-binary-handling-mode[binary handling mode] setting
```
Comment (jpechane) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc:1124]: I wonder if this should be inlided as is now or whether the better approach would be like with numeric and temporal values. So extracting them into a separate section and show the data mapping based on the config property setting.

Comment (gunnarmorling) [reply documentation/modules/ROOT/pages/connectors/postgresql.adoc:1124]: I was contemplating about it, too. But I don't think it's worth it for this single one. Perhaps for MySQL, but actually I'd prefer to have one single table with all the props, i.e. converge on this style added here.

**Comment by gunnarmorling** [comment]
Thanks for reviewing, @hashhar!

> the current switching on BinaryHandlingMode doesn't allow for custom implementations but it shouldn't be an issue since I personally can't think of any other conversion I would've done apart from the ones already implemented.

Yeah, I don't think that's needed either.

> Custom may be useful if somebody has added their own custom data types into PostgreSQL and can implement a matching Java type to return as but I really hope nobody went to those lengths.

In that case they should be using the converter API that was recently added. The "unknown types" option is generally a measure of last resort, i.e. people should preferably use converters and/or send PRs for adding more type support to Debezium proper.

**Comment by gunnarmorling** [comment]
Rebased and force-pushed.

**Comment by jpechane** [comment]
@gunnarmorling Last question, as this is a common connector feature why is it unavalibale to other connectors like SQL Server?

**Comment by gunnarmorling** [comment]
> why is it unavalibale to other connectors like SQL Server?

Good point, largely a question of getting done with it. I've logged https://issues.redhat.com/browse/DBZ-2155 as a follow-up.

**Comment by jpechane** [comment]
@jfinzel @gunnarmorling Applied, thanks a lot!

**Pull request merged**